core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1238,14 +1238,22 @@ class KafkaApis(val requestChannel: RequestChannel,
   }
 
   def handleListGroupsRequest(request: RequestChannel.Request) {
-    if (!authorize(request.session, Describe, Resource.ClusterResource)) {
+    val (error, groups) = groupCoordinator.handleListGroups()
+    // the alternative where at least some Describe Group ACLs for current principal is needed for running ListGroups
+    if (authorize(request.session, Describe, Resource.ClusterResource))
       sendResponseMaybeThrottle(request, requestThrottleMs =>
-        request.body[ListGroupsRequest].getErrorResponse(requestThrottleMs, Errors.CLUSTER_AUTHORIZATION_FAILED.exception))
-    } else {
-      val (error, groups) = groupCoordinator.handleListGroups()
-      val allGroups = groups.map { group => new ListGroupsResponse.Group(group.groupId, group.protocolType) }
-      sendResponseMaybeThrottle(request, requestThrottleMs =>
-        new ListGroupsResponse(requestThrottleMs, error, allGroups.asJava))
+        new ListGroupsResponse(requestThrottleMs, error, groups.map { group => new ListGroupsResponse.Group(group.groupId, group.protocolType) }.asJava))
+    else {
+      val principalGroupAcls: List[Acl] = authorizer.map(_.getAcls().filter(_._1.resourceType == Group).values.flatten.toList).toList.flatten
+        .filter(acl => acl.principal == request.session.principal)
+      if (principalGroupAcls.isEmpty)
+        sendResponseMaybeThrottle(request, requestThrottleMs =>
+          request.body[ListGroupsRequest].getErrorResponse(requestThrottleMs, Errors.CLUSTER_AUTHORIZATION_FAILED.exception))
+      else {
+        val filteredGroups = groups.filter(group => authorize(request.session, Describe, new Resource(Group, group.groupId, LITERAL)))
```

Comment:
Can we just return CLUSTER_AUTHORIZATION_FAILED error, if filteredGroups is empty?  This may avoid getAcls().filter iteration above.

Reply:
Yes, I think that works, and should simply the error condition. Updated the PR. Thanks!